### PR TITLE
Upgrade isort to 4.3.4.

### PR DESCRIPTION
**Pull request by jsirois**
This adds support for more python3 stdlib symbols.

Also introduce a build script.

I found these instructions in the pants repo late in the game: https://github.com/pantsbuild/pants/blob/master/src/python/pants/backend/python/tasks/python_isort.py#L25-L29

In practice they do not look necessary. I was able to do the following to get my local pants running on isort 4.3.4:
```
binaries $ ./build_isort.sh 4.3.4
binaries $ mkdir -p ~/.cache/pants/bin/isort/4.3.4
binaries $ cp build-support/scripts/isort/4.3.4/isort.pex ~/.cache/pants/bin/isort/4.3.4/isort.pex
binaries $ cd ~/dev/pantsbuild/pants
pants $ vi src/python/pants/base/parse_context.py
pants $ git diff
diff --git a/src/python/pants/base/parse_context.py b/src/python/pants/base/parse_context.py
index 278467514..4af400ee1 100644
--- a/src/python/pants/base/parse_context.py
+++ b/src/python/pants/base/parse_context.py
@@ -5,6 +5,7 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
 
 import functools
+
 import threading
 from builtins import object
 
pants $ PANTS_ISORT_VERSION=4.3.4 ./build-support/bin/isort.sh -f
Fixing /home/jsirois/dev/pantsbuild/jsirois-pants/src/python/pants/base/parse_context.py

pants $
```

### CODE REVIEW

**Comment by jsirois** [comment]
N.B. This is a bit of a straw man. As noted, I found the instructions in the pants repo late after I had started down the path of making a binary like any other over here. I'm not convinced either approach is better, but perhaps the 'TODO: Explain why we don't invoke isort directly.' comment in the `IsortPythonTask` class docs can be explained out loud.

**Comment by wisechengyi** [comment]
Hey @jsirois sorry I am responsible for `TODO: Explain why we don't invoke isort directly.` In the end the task was made up in hackweek :P

isort for python2 has import side effect breaking pants' run_tracker's hook
https://github.com/timothycrosley/isort/issues/456

but it might work with python3 as noted in the ticket.

File: build_pex.bash [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    mkdir -p "$(dirname "${dir}")"
+    (
+      cd "$(ensure_venv_installed)"
+      python2.7 virtualenv.py "${dir}"
+    )
+  fi 
+  source "${activate}"
+}
+
+PEX_VERSION="1.4.4"
```
Comment (kwlzn) [inline build_pex.bash]: `pex` has long published itself as a pex - on the github releases page at e.g. `https://github.com/pantsbuild/pex/releases/download/v1.4.4/pex27`.

any reason to not just curl this single file bin at `$PEX_VERSION`, chmod and run vs setting up a venv just to get `pex(1)`?

**Comment by jsirois** [comment]
Aha. I read invoke directly as build a pex directly which is what this PR does. What then is this for?:
+ https://github.com/pantsbuild/pants/blob/master/contrib/python/src/python/pants/contrib/python/isort/isort.py
+ https://github.com/pantsbuild/pants/blob/master/contrib/python/src/python/pants/contrib/python/isort/BUILD

File: build_pex.bash (hunk 1 above)
Comment (jsirois) [reply build_pex.bash]: Nope! Being real dumb since I knew that / have done a handful of those releases. Will fix.

**Comment by wisechengyi** [comment]
isort.py is the same as the script used to invoke isort after a `pip install isort`.
```
$ cat /opt/example/bin/isort
#!/opt/example/opt/python/bin/python2.7

# -*- coding: utf-8 -*-
import re
import sys

from isort.main import main

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
    sys.exit(main())
```

> build a pex directly which is what this PR does

Yeah what you did should work too. 

Please let me know if anything else is unclear.

File: build_pex.bash [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+function run_pex() {
+  local -r version="${PEX_VERSION}"
+
+  local -r pex_dir="${CACHE_ROOT}/pex/${version}"
+  activate_venv "${pex_dir}"
+  if [[ ! -x "${pex_dir}/bin/pex" ]]; then
+    pip install "pex==${version}" requests
```
Comment (wisechengyi) [inline build_pex.bash]: is `requests` needed?

Comment (kwlzn) [reply build_pex.bash]: `requests` is currently optional for pex - it will use it if available, but doesn't concretely depend on it.

Comment (wisechengyi) [reply build_pex.bash]: ah ok. just making sure `requests` won't end up in `isort.pex`, which seems to be separate here.

Comment (kwlzn) [reply build_pex.bash]: we pre-bundle `requests` into the release published `pex` pex - so it'd only end up in that toolchain pex and not in the `isort.pex` we will later construct.

**Comment by stuhood** [comment]
(looks like you've got reviews for the rest, so will ignore)

But as to building a pex for `isort`: one thing that we could certainly do is have pants build an isort pex dynamically, the same way we do for `pytest`. I've suggested this to Chris as a possible path forward for python linting in general.

**Comment by jsirois** [comment]
> we could certainly do is have pants build an isort pex dynamically,

Absolutely - this was a confusing adventure and it would be great to bootstrap tools for languages pants understands via pants. I'll take a look at that.

**Pull request merged**